staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply.go

```diff
@@ -0,0 +1,150 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handlers
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apply"
+	"k8s.io/apimachinery/pkg/apply/parse"
+	"k8s.io/apimachinery/pkg/apply/strategy"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+type applyPatcher struct {
+	*patcher
+
+	model proto.Schema
+}
+
+// TODO(apelisse): workflowId needs to be passed as a query
+// param/header, and a better defaulting needs to be defined too.
+const workflowId = "default"
+
+func (p *applyPatcher) convertCurrentVersion(obj runtime.Object) (map[string]interface{}, error) {
+	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.kind.GroupVersion())
+	if err != nil {
+		return nil, err
+	}
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
+}
+
+func (p *applyPatcher) extractLastIntent(obj runtime.Object, workflow string) (map[string]interface{}, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
+	}
+	last := make(map[string]interface{})
+	if accessor.GetLastApplied()[workflow] != "" {
+		if err := json.Unmarshal([]byte(accessor.GetLastApplied()[workflow]), &last); err != nil {
+			return nil, fmt.Errorf("couldn't unmarshal last applied field: %v", err)
+		}
+	}
+	return last, nil
+}
+
+func (p *applyPatcher) getNewIntent() (map[string]interface{}, error) {
+	patch := make(map[string]interface{})
+	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
+	}
+	return patch, nil
+}
+
+func (p *applyPatcher) convertResultToUnversioned(result apply.Result) (runtime.Object, error) {
+	voutput, err := p.creater.New(p.kind)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create empty output object: %v", err)
+	}
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.MergedResult.(map[string]interface{}), voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result back: %v", err)
+	}
+	p.defaulter.Default(voutput)
+
+	uoutput, err := p.toUnversioned(voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to unversioned: %v", err)
+	}
+
+	return uoutput, nil
+}
+
+func (p *applyPatcher) saveNewIntent(workflow string, dst runtime.Object) error {
+	// We want to consistently save the intent in JSon.
+	json, err := yaml.YAMLToJSON(p.patchBytes)
+	if err != nil {
+		return fmt.Errorf("failed to convert patch to JSon: %v", err)
+	}
+	accessor, err := meta.Accessor(dst)
+	if err != nil {
+		return fmt.Errorf("couldn't get accessor: %v", err)
+	}
+	m := accessor.GetLastApplied()
```

Comment:
This (i.e., modifying a map that's embedded in an object) is safe only because we know we just constructed this object. Maybe worth a note to the reader (in a follow up)?

Reply:
I can change now.